resource-agents: vSphere K8s cluster provider

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
Resolves https://github.com/bottlerocket-os/bottlerocket-test-system/issues/35


**Description of changes:**

```
    vsphere-vm-resource: refactor out functions for shared use
    
    This refactors out some functions needed for implementing the vSphere K8s
    cluster provider so they can be shared between the two crates

```
```
    bottlerocket-agents: new 'vsphere-k8s-cluster-provider' agent
    
    This adds a new resource agent for provisioning vSphere K8s clusters via
    EKS-A.

```

**Testing done:**
Applied the following resource spec for a vSphere K8s cluster:
```
apiVersion: testsys.bottlerocket.aws/v1
kind: Resource
metadata:
  name: vsphere-k8s-cluster-1-23
  namespace: testsys-bottlerocket-aws
spec:
  agent:
    name: vsphere-k8s-cluster-resource-agent
    image: <>.dkr.ecr.us-west-2.amazonaws.com/vsphere-k8s-cluster-resource-agent:latest
    keepRunning: false
    privileged: true
    timeout: 20d
    secrets:
      vsphereCredentials: <>
    configuration:
      name: br-eksa-123
      controlPlaneEndpointIp: <>
      creation_policy: IfNotExists
      version: v1.23
      ovaName: "bottlerocket-vmware-k8s-1.23-x86_64-v1.10.1.ova"
      tufRepo:
        metadataUrl: "https://updates.bottlerocket.aws/2020-07-07/vmware-k8s-1.23/x86_64/"
        targetsUrl: "https://updates.bottlerocket.aws/targets"
      vcenterHostUrl: <>
      vcenterDatacenter: <>
      vcenterDatastore: <>
      vcenterNetwork: <>
      vcenterResourcePool: <>
      vcenterWorkloadFolder: <>
      mgmtClusterKubeconfigBase64: <>
      destructionPolicy: OnDeletion

```

Creation pod runs to completion without problem:
<details>

```bash
$ kubectl logs -f vsphere-k8s-clu70916266-4155-478e-9689-2b5dcead3c56-creatiwnkx6 -n testsys-bottlerocket-aws
[2022-10-21T18:53:07Z INFO  resource_agent::agent] Initializing Agent
[2022-10-21T18:53:07Z INFO  vsphere_k8s_cluster_resource_agent::vsphere_k8s_cluster_provider] Creation policy is 'Create' and cluster 'br-eksa-123' does not exist: creating cluster
[2022-10-21T18:53:08Z INFO  vsphere_k8s_cluster_resource_agent::vsphere_k8s_cluster_provider] Downloading OVA 'bottlerocket-vmware-k8s-1.23-x86_64-v1.10.1.ova'
[2022-10-21T18:53:10Z INFO  vsphere_k8s_cluster_resource_agent::vsphere_k8s_cluster_provider] Importing OVA and creating a VM template out of it
[2022-10-21T18:53:22Z INFO  vsphere_k8s_cluster_resource_agent::vsphere_k8s_cluster_provider] Tagging VM template
2022-10-21T18:53:49.895Z        V4      Logger init completed   {"vlevel": 4}
2022-10-21T18:53:50.012Z        V4      Reading bundles manifest        {"url": "https://anywhere-assets.eks.amazonaws.com/releases/bundles/18/manifest.yaml"}
2022-10-21T18:53:50.072Z        V4      Relative network path specified, using path /SDDC-Datacenter/network/sddc-cgw-network-2
2022-10-21T18:53:50.131Z        V2      Pulling docker image    {"image": "public.ecr.aws/eks-anywhere/cli-tools:v0.11.4-eks-a-18"}
2022-10-21T18:53:56.618Z        V3      Initializing long running container     {"name": "eksa_1666378430131231140", "image": "public.ecr.aws/eks-anywhere/cli-tools:v0.11.4-eks-a-18"}
2022-10-21T18:53:58.588Z        V4      Task start      {"task_name": "setup-validate"}
2022-10-21T18:53:58.588Z        V0      Performing setup and validations
2022-10-21T18:53:58.588Z        V4      Relative network path specified, using path /SDDC-Datacenter/network/sddc-cgw-network-2
2022-10-21T18:53:58.605Z        V0      ✅ Connected to server
2022-10-21T18:53:58.881Z        V0      ✅ Authenticated to vSphere
2022-10-21T18:53:59.481Z        V0      ✅ Datacenter validated
2022-10-21T18:53:59.786Z        V0      ✅ Network validated
2022-10-21T18:53:59.786Z        V1      SSHUsername is not set or is empty for VSphereMachineConfig, using default      {"machineConfig": "br-eksa-123-node", "user": "ec2-user"}
2022-10-21T18:54:00.363Z        V0      Warning: Your VM template has no snapshots. Defaulting to FullClone mode. VM provisioning might take longer.
2022-10-21T18:54:00.648Z        V0      ✅ Datastore validated
2022-10-21T18:54:00.935Z        V0      ✅ Folder validated
2022-10-21T18:54:01.224Z        V0      ✅ Resource pool validated
2022-10-21T18:54:02.329Z        V0      ✅ Control plane and Workload templates validated
2022-10-21T18:54:03.190Z        V0      Provided control plane sshAuthorizedKey is not set or is empty, auto-generating new key pair...
2022-10-21T18:54:04.517Z        V0      Private key saved to br-eksa-123/eks-a-id_rsa. Use 'ssh -i br-eksa-123/eks-a-id_rsa <username>@<Node-IP-Address>' to login to your cluster node
2022-10-21T18:54:16.033Z        V0      ✅ cloudadmin@vmc.local user vSphere privileges validated
2022-10-21T18:54:16.033Z        V0      ✅ Vsphere Provider setup is valid
2022-10-21T18:54:17.184Z        V0      ✅ Validate certificate for registry mirror
2022-10-21T18:54:17.184Z        V0      ✅ Validate authentication for git provider
2022-10-21T18:54:17.184Z        V0      ✅ Validate cluster name
2022-10-21T18:54:17.184Z        V0      ✅ Validate gitops
2022-10-21T18:54:17.184Z        V0      ✅ Validate identity providers' name
2022-10-21T18:54:17.184Z        V0      ✅ Validate management cluster has eksa crds
2022-10-21T18:54:17.184Z        V0      ✅ Create preflight validations pass
2022-10-21T18:54:17.184Z        V4      Task finished   {"task_name": "setup-validate", "duration": "18.595666711s"}
2022-10-21T18:54:17.184Z        V4      ----------------------------------
2022-10-21T18:54:17.184Z        V4      Task start      {"task_name": "bootstrap-cluster-init"}
2022-10-21T18:54:17.184Z        V4      Task finished   {"task_name": "bootstrap-cluster-init", "duration": "2.243µs"}
2022-10-21T18:54:17.184Z        V4      ----------------------------------
2022-10-21T18:54:17.184Z        V4      Task start      {"task_name": "workload-cluster-init"}
2022-10-21T18:54:17.184Z        V0      Creating new workload cluster
2022-10-21T18:54:18.398Z        V3      Waiting for external etcd to be ready   {"cluster": "br-eksa-123"}
2022-10-21T18:56:18.609Z        V3      External etcd is ready
2022-10-21T18:56:18.609Z        V3      Waiting for control plane to be ready
2022-10-21T18:58:48.500Z        V3      Waiting for workload kubeconfig generation      {"cluster": "br-eksa-123"}
2022-10-21T18:58:49.060Z        V3      Waiting for controlplane and worker machines to be ready
2022-10-21T18:58:49.365Z        V4      Nodes are not ready yet {"total": 3, "ready": 2, "cluster name": "br-eksa-123"}
2022-10-21T18:58:49.672Z        V4      Nodes are not ready yet {"total": 3, "ready": 2, "cluster name": "br-eksa-123"}
2022-10-21T18:58:51.001Z        V4      Nodes ready     {"total": 3}
2022-10-21T18:58:51.001Z        V0      Installing networking on workload cluster
2022-10-21T18:58:52.910Z        V0      Installing storage class on cluster
2022-10-21T18:58:53.290Z        V4      Installing machine health checks on bootstrap cluster
2022-10-21T18:58:53.809Z        V4      Task finished   {"task_name": "workload-cluster-init", "duration": "4m36.62540685s"}
2022-10-21T18:58:53.809Z        V4      ----------------------------------
2022-10-21T18:58:53.809Z        V4      Task start      {"task_name": "install-resources-on-management-cluster"}
2022-10-21T18:58:53.809Z        V4      Task finished   {"task_name": "install-resources-on-management-cluster", "duration": "1.816µs"}
2022-10-21T18:58:53.809Z        V4      ----------------------------------
2022-10-21T18:58:53.809Z        V4      Task start      {"task_name": "capi-management-move"}
2022-10-21T18:58:53.809Z        V4      Task finished   {"task_name": "capi-management-move", "duration": "834ns"}
2022-10-21T18:58:53.809Z        V4      ----------------------------------
2022-10-21T18:58:53.809Z        V4      Task start      {"task_name": "eksa-components-install"}
2022-10-21T18:58:53.809Z        V0      Creating EKS-A CRDs instances on workload cluster
2022-10-21T18:58:53.812Z        V4      Applying eksa yaml resources to cluster
2022-10-21T18:58:54.339Z        V1      Applying Bundles to cluster
2022-10-21T18:58:55.001Z        V4      Applying eksd manifest to cluster
2022-10-21T18:58:57.201Z        V4      Task finished   {"task_name": "eksa-components-install", "duration": "3.391983322s"}
2022-10-21T18:58:57.201Z        V4      ----------------------------------
2022-10-21T18:58:57.201Z        V4      Task start      {"task_name": "gitops-manager-install"}
2022-10-21T18:58:57.201Z        V0      Installing GitOps Toolkit on workload cluster
2022-10-21T18:58:57.201Z        V0      GitOps field not specified, bootstrap flux skipped
2022-10-21T18:58:57.201Z        V4      Task finished   {"task_name": "gitops-manager-install", "duration": "11.854µs"}
2022-10-21T18:58:57.201Z        V4      ----------------------------------
2022-10-21T18:58:57.201Z        V4      Task start      {"task_name": "write-cluster-config"}
2022-10-21T18:58:57.201Z        V0      Writing cluster config file
2022-10-21T18:58:57.203Z        V4      Task finished   {"task_name": "write-cluster-config", "duration": "2.052509ms"}
2022-10-21T18:58:57.203Z        V4      ----------------------------------
2022-10-21T18:58:57.203Z        V4      Task start      {"task_name": "delete-kind-cluster"}
2022-10-21T18:58:57.203Z        V0      🎉 Cluster created!
2022-10-21T18:58:57.203Z        V4      Task finished   {"task_name": "delete-kind-cluster", "duration": "7.66µs"}
2022-10-21T18:58:57.203Z        V4      ----------------------------------
2022-10-21T18:58:57.203Z        V4      Task start      {"task_name": "install-curated-packages"}
--------------------------------------------------------------------------------------
The Amazon EKS Anywhere Curated Packages are only available to customers with the
Amazon EKS Anywhere Enterprise Subscription
--------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------
Curated packages cannot be installed as cert-manager is not present in the cluster.
This is most likely caused by an action to install curated packages at a workload
cluster. Refer to https://anywhere.eks.amazonaws.com/docs/tasks/troubleshoot/packages/
for how to resolve this issue.
--------------------------------------------------------------------------------------
2022-10-21T18:58:57.494Z        V0      ❌ Curated Packages Installation Failed...
2022-10-21T18:58:57.494Z        V4      Task finished   {"task_name": "install-curated-packages", "duration": "290.286471ms"}
2022-10-21T18:58:57.494Z        V4      ----------------------------------
2022-10-21T18:58:57.494Z        V4      Tasks completed {"duration": "4m58.905757983s"}
2022-10-21T18:58:57.494Z        V3      Logging out from current govc session
2022-10-21T18:58:58.119Z        V3      Cleaning up long running container      {"name": "eksa_1666378430131231140"}
[2022-10-21T18:58:58Z INFO  vsphere_k8s_cluster_resource_agent::vsphere_k8s_cluster_provider] Scaling default NodeGroup machinedeployments replicas to 0 machinedeployment.cluster.x-k8s.io/br-eksa-123-md-0 scaled
```

</details>

Destruction runs to completion without problem after deleting the resource:
<details>

```bash
$ kubectl logs -f vsphere-k8s-clu70916266-4155-478e-9689-2b5dcead3c56-destruktfkr -n testsys-bottlerocket-aws
[2022-10-21T18:59:57Z INFO  resource_agent::agent] Initializing Agent
2022-10-21T18:59:58.418Z        V4      Logger init completed   {"vlevel": 4}
2022-10-21T18:59:58.533Z        V4      Reading bundles manifest        {"url": "https://anywhere-assets.eks.amazonaws.com/releases/bundles/18/manifest.yaml"}
2022-10-21T18:59:58.562Z        V4      Relative network path specified, using path /SDDC-Datacenter/network/sddc-cgw-network-2
2022-10-21T18:59:58.611Z        V2      Pulling docker image    {"image": "public.ecr.aws/eks-anywhere/cli-tools:v0.11.4-eks-a-18"}
2022-10-21T19:00:04.009Z        V3      Initializing long running container     {"name": "eksa_1666378798611379528", "image": "public.ecr.aws/eks-anywhere/cli-tools:v0.11.4-eks-a-18"}
2022-10-21T19:00:06.077Z        V4      Task start      {"task_name": "setup-and-validate"}
2022-10-21T19:00:06.077Z        V0      Performing provider setup and validations
2022-10-21T19:00:06.077Z        V4      Task finished   {"task_name": "setup-and-validate", "duration": "33.101µs"}
2022-10-21T19:00:06.077Z        V4      ----------------------------------
2022-10-21T19:00:06.077Z        V4      Task start      {"task_name": "management-cluster-init"}
2022-10-21T19:00:06.077Z        V4      Task finished   {"task_name": "management-cluster-init", "duration": "1.049µs"}
2022-10-21T19:00:06.077Z        V4      ----------------------------------
2022-10-21T19:00:06.077Z        V4      Task start      {"task_name": "delete-workload-cluster"}
2022-10-21T19:00:06.077Z        V0      Deleting workload cluster
2022-10-21T19:00:33.606Z        V4      Task finished   {"task_name": "delete-workload-cluster", "duration": "27.528800824s"}
2022-10-21T19:00:33.606Z        V4      ----------------------------------
2022-10-21T19:00:33.606Z        V4      Task start      {"task_name": "clean-up-git-repo"}
2022-10-21T19:00:33.606Z        V0      Clean up Git Repo
2022-10-21T19:00:33.606Z        V0      GitOps field not specified, clean up git repo skipped
2022-10-21T19:00:33.606Z        V4      Task finished   {"task_name": "clean-up-git-repo", "duration": "9.966µs"}
2022-10-21T19:00:33.606Z        V4      ----------------------------------
2022-10-21T19:00:33.606Z        V4      Task start      {"task_name": "kind-cluster-delete"}
2022-10-21T19:00:33.606Z        V0      Bootstrap cluster information missing - skipping delete kind cluster
2022-10-21T19:00:33.606Z        V0      🎉 Cluster deleted!
2022-10-21T19:00:33.606Z        V4      Task finished   {"task_name": "kind-cluster-delete", "duration": "9.076µs"}
2022-10-21T19:00:33.606Z        V4      ----------------------------------
2022-10-21T19:00:33.606Z        V4      Tasks completed {"duration": "27.528983342s"}
2022-10-21T19:00:33.606Z        V3      Logging out from current govc session
2022-10-21T19:00:34.198Z        V3      Cleaning up long running container      {"name": "eksa_1666378798611379528"}
rpc error: code = NotFound desc = an error occurred when try to find container "b0624db8aa5f5349c38be241495784d597bd56e975fdc4064b635dbfaa782c50": not found

```
</details>

With `testsys run vmware`, the cluster can be used to run migration tests/conformance tests:

<details>

```bash
$ testsys run vmware \
   --cluster-endpoint 198.19.11.123 \
   --cluster-name br-eksa-123 \
   --target-cluster-kubeconfig-path br-eksa-123-eks-a-cluster.kubeconfig \
   --test-agent-image <>.dkr.ecr.us-west-2.amazonaws.com/sonobuoy-test-agent:latest \
   --network sddc-cgw-network-2 \
   --name vmware-123-quick \
   --ova-name bottlerocket-vmware-k8s-1.23-x86_64-v1.10.0.ova \
   --vcenter-url <> \
   --vm-count 2 \
   --vm-provider-image <>.dkr.ecr.us-west-2.amazonaws.com/vsphere-vm-resource-agent:latest \
   --vsphere-secret vcentercreds \
   --workload-folder etung \
   --upgrade-downgrade \
   --migration-agent-image <>.dkr.ecr.us-west-2.amazonaws.com/migration-test-agent:latest \
   --starting-version v1.10.0 \
   --upgrade-version v1.10.1 \
   --tuf-repo-metadata-url https://updates.bottlerocket.aws/2020-07-07/vmware-k8s-1.22/x86_64/ \
   --tuf-repo-targets-url https://updates.bottlerocket.aws/targets/
Created resource object 'br-eksa-123-vms'
Created test object 'vmware-123-quick-1-initial'
Created test object 'vmware-123-quick-2-migrate'
Created test object 'vmware-123-quick-3-migrated'
Created test object 'vmware-123-quick-4-migrate'
Created test object 'vmware-123-quick-5-final'
$ testsys status -c -r
 NAME                          TYPE         STATE       PASSED   SKIPPED   FAILED 
 Controller                    Controller   Running                               
 br-eksa-123-vms               Resource     completed                             
 vmware-123-quick-1-initial    Test         passed      1        7049      0      
 vmware-123-quick-2-migrate    Test         passed      2        0         0      
 vmware-123-quick-3-migrated   Test         passed      1        7049      0      
 vmware-123-quick-4-migrate    Test         passed      2        0         0      
 vmware-123-quick-5-final      Test         passed      1        7049      0      
 vsphere-k8s-cluster-1-23      Resource     completed 
```

</details>

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten committed:
vsphere-vm-resource: refactor out functions for shared use

This refactors out some functions needed for implementing the vSphere
K8s cluster provider so they can be shared between the two crates

etungsten committed:
bottlerocket-agents: new 'vsphere-k8s-cluster-provider' agent

This adds a new resource agent for provisioning vSphere K8s clusters via
EKS-A.

etungsten commented:
The "Build/Images" check is failing because I changed the bottlerocket test tools image Dockerfile and the check is using the default tools image in the ECR repository.

etungsten commented:
Taking this to draft until https://github.com/bottlerocket-os/bottlerocket-test-system/pull/615 merges and we push out the new tools image.

webern commented:
>  the check is using the default tools image in the ECR repository

Really? You mean latest? That's not right.

webern commented on bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs:
```diff
@@ -0,0 +1,763 @@
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ProductionMemo {
+    /// In this resource we put some traces here that describe what our provider is doing.
```
nice idea

webern commented on bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs:
```diff
@@ -0,0 +1,763 @@
+        let mgmt_kubeconfig_path = format!("{}/mgmt.kubeconfig", WORKING_DIR);
+        let encoded_kubeconfig = if do_create {
+            // Check CAPI management cluster is accessible and valid
+            debug!("Decoding and writing out kubeconfig for the CAPI management cluster");
```
This agent requires a CAPI management cluster to already exist?

etungsten replied:
Yes, like you mentioned in the review message, I wanted to avoid bootstrapping with kind since the docker daemon is running with `vfs` as the underlying storage driver and it's incredibly inefficient since each container layer is its own directory. And setting up a kind bootstrap cluster requires a lot of containers to be running. I can try again to see if this is acceptable though.

Setting up overlayfs requires installing docker on the testsys node host and mounting in `/var/lib/docker` into the dind container which I find unappetizing as well.

By having an external management cluster, I don't have to worry about all of that. In addition, the management cluster becomes a source of truth for the state of all provisioned clusters and I can retrieve a kubeconfig for any provisioned cluster as long as I have access to the mgmt cluster. There are cons with this approach though. 

- I believe this makes integrating with `cargo make test` harder since the resource agent is no longer self-contained and requires an external kubeconfig (@ecpullen probably has opinions on this). 
- Someone needs to maintain the management cluster and remember to upgrade the CAPI, CAPV CRDs when EKS-A bumps versions.

webern replied:
Makes sense. I think this is a great solution and we can consider "Don't require an external CAPI cluster" to be a separate issue/feature-request.

ecpullen replied:
The additional kubeconfig will either need to have a know location in the monorepo (like `testsys.kubeconfig`) or we will have to include it (as a string) in the `Test.toml` file. I prefer the former option, but not having to worry about it would be nice.

etungsten replied:
I tried without a external management cluster and let EKS-A set up the bootstrap cluster in kind (kind running in docker that's running in docker btw lol) and it failed due to networking issues:
```
2022-10-25T20:35:28.490Z        V4      ----------------------------------
2022-10-25T20:35:28.490Z        V4      Task start      {"task_name": "bootstrap-cluster-init"}
2022-10-25T20:35:28.490Z        V0      Creating new bootstrap cluster
2022-10-25T20:35:28.491Z        V4      Creating kind cluster   {"name": "br-eksa-122-eks-a-cluster", "kubeconfig": "br-eksa-122/generated/br-eksa-122.kind.kubeconfig"}
2022-10-25T20:35:28.491Z        V6      Executing command       {"cmd": "/usr/bin/docker exec -i eksa_1666730106969714044 kind create cluster --name br-eksa-122-eks-a-cluster --kubeconfig br-eksa-122/generated/br-eksa-122.kind.kubeconfig --image public.ecr.aws/eks-anywhere/kubernetes-sigs/kind/node:v1.22.15-eks-d-1-22-11-eks-a-19 --config br-eksa-122/generated/kind_tmp.yaml"}
ERRO[2022-10-25T20:35:53.428755287Z] Could not add route to IPv6 network fc00:f853:ccd:e793::1/64 via device br-3443489d9de0: network is down 
time="2022-10-25T20:35:56.630430522Z" level=info msg="loading plugin \"io.containerd.event.v1.publisher\"..." runtime=io.containerd.runc.v2 type=io.containerd.event.v1
time="2022-10-25T20:35:56.630486720Z" level=info msg="loading plugin \"io.containerd.internal.v1.shutdown\"..." runtime=io.containerd.runc.v2 type=io.containerd.internal.v1
time="2022-10-25T20:35:56.630493149Z" level=info msg="loading plugin \"io.containerd.ttrpc.v1.task\"..." runtime=io.containerd.runc.v2 type=io.containerd.ttrpc.v1
time="2022-10-25T20:35:56.630582130Z" level=info msg="starting signal loop" namespace=moby path=/run/docker/containerd/daemon/io.containerd.runtime.v2.task/moby/99c8a19b51e1e85305f365229030c3b6b7e58f7423f120b88f02d0162ee1de7e pid=2752 runtime=io.containerd.runc.v2
```

So unless we want to invest time investigating how to get kind to work under 3 layers of container abstraction, I think we should stick with external mgmt cluster for now.

webern replied:
:turtle: :turtle: :turtle: :turtle: :turtle:

webern commented on bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs:
```diff
@@ -0,0 +1,763 @@
+            .context(resources, "Error sending cluster creation message")?;
+
+        let mgmt_kubeconfig_path = format!("{}/mgmt.kubeconfig", WORKING_DIR);
+        let encoded_kubeconfig = if do_create {
```
This is a very large if block. Is this practical? (sometimes this would be impractical if too many arguments and returns are required)

```rust
if do_create {
   create_cluster(x,y,z)?
} else {
  ////
}
```
```

etungsten replied:
I already tried slimming it down with `create_vsphere_k8s_cluster`. I can maybe slim it down further by capturing the preflight/postflight stuff in other functions.

webern commented on bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs:
```diff
@@ -0,0 +1,763 @@
+    creation_policy: CreationPolicy,
+) -> ProviderResult<(bool, String)> {
+    let cluster_exists = does_cluster_exist(cluster_info).await?;
+    match creation_policy {
```
This torturous function (my fault) is duplicated, I think. I wonder if it would make sense to take `cluster_exists` as input to this function and then the EKS and ECS agents could use the same function.

etungsten replied:
I was wondering the same. I'll try refactoring this.

webern commented on bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs:
```diff
@@ -0,0 +1,763 @@
+    let ova_name = config.ova_name.to_owned();
+    info!("Downloading OVA '{}'", &config.ova_name);
+    let outdir = Path::new("/local/");
+    tokio::task::spawn_blocking(move || -> ProviderResult<()> {
```
are you doing this because download_target is a blocking function? Which it is because `tough` is not async? https://github.com/awslabs/tough/issues/213

Does `spawn_blocking` created a dedicated thread? It would be safest if it does because of the multiple tokio runtimes issue that reqwest::blocking can cause.

etungsten replied:
> Does spawn_blocking created a dedicated thread?

Yes, it creates a thread where blocking is ok.

webern commented on bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs:
```diff
@@ -0,0 +1,763 @@
+                ),
+            ));
+        }
+        sleep(Duration::from_millis(200));
```
suspicious. You probably want to call out why this is here and how arbitrary/non-arbitrary 200 ms is.

etungsten replied:
I didn't want to overload the vsphere API with create/attach tag calls. The 200ms is completely ~unnecessary~ arbituary. I'll experiment with his a bit more and add a comment.

etungsten replied:
I typo'd "arbitrary" into "unnecessary". Thanks brain.
Going to try without the sleep and test.

webern replied:
It's fine to have it if it makes you feel more confident in reliability. You can just a comment to say why you think it might be helpful. No need to be overly scientific with the value.

webern reviewed (commented):
This is great work and exciting that we will have this functionality.

My main question is about requiring a CAPI cluster to pre-exist. I think I understand the choice... otherwise you would need a kind cluster running "docker in docker" inside this container, right? And maybe that feels just too horrible?

The other feedback I have is that `vsphere_k8s_cluster_provider.rs` is basically a big long procedural script. My person preference would be to look for a bunch of places where you could refactor out some functions (even if they are only being called once) just for the sake of readability.

etungsten commented on bottlerocket/agents/src/bin/vsphere-vm-resource-agent/vsphere_vm_provider.rs:
```diff
@@ -136,15 +136,14 @@ impl Create for VMCreator {
-            Resources::Clear,
-            "Failed to write out kubeconfig for vSphere cluster",
-        )?;
+        if let Some(kubeconfig_base64) = vsphere_cluster.kubeconfig_base64 {
```
I probably shouldn't share the `VSphereK8sClusterInfo` struct definition between the vm resource agent and the cluster provider agent since the vm resource agent should never be invoked if the kubeconfig for the cluster is missing.

That should be enforced by the spec. By making the `kubeconfig_base64` field optional, we lose that validation.

etungsten committed:
bottlerocket-agents: refactor cluster creation policy processing

This refactors out the duplicated logic between eks cluster provider and
vsphere k8s cluster cluster provider for determining whether cluster
creation is necessary according to the CreationPolicy

etungsten commented:
Push above addresses @webern 's comments.

- Removes the sleeps in the loop for creating and attaching tags to VM templates
- Refactor `fn create_cluster` to make it easier to read
- Refactor out `is_cluster_creation_required` to remove duplication between EKS resource agent and vSphere K8s cluster resource agent

Repeated testing and the results are the same as before as described in the PR description

etungsten commented:
Push above rebases onto develop and fixes conflicts in EKS resource agent

webern reviewed (approved):
nice

ecpullen reviewed (commented):
Can you add a tutorial for using this agent somewhere?

etungsten commented:
Push above adds a README for the vSphere K8s cluster resource agent on how to configure the agent and more info on how to set up the initial management cluster.

etungsten commented:
Push above rebases onto develop and fixes conflicts with https://github.com/bottlerocket-os/bottlerocket-test-system/pull/537

etungsten commented:
Push above addresses lints caught by `mdlint` during Lint github action.

ecpullen commented on bottlerocket/agents/src/lib.rs:
```diff
@@ -6,6 +6,51 @@ This `lib.rs` provides code that is used by multiple agent binaries or used by t
+pub mod vsphere;
+
+/// Determines whether a cluster resource needs to be created given its creation policy
+pub async fn is_cluster_creation_required(
+    cluster_exists: &bool,
+    cluster_name: &str,
+    creation_policy: &CreationPolicy,
+) -> ProviderResult<(bool, String)> {
+    match creation_policy {
+        CreationPolicy::Create if *cluster_exists =>
+            Err(
+                ProviderError::new_with_context(
+                    Resources::Clear, format!(
+                        "The cluster '{}' already existed and creation policy '{:?}' requires that it not exist",
+                        cluster_name,
+                        creation_policy
+                    )
+                )
+            ),
+        CreationPolicy::Never if !*cluster_exists =>
+            Err(
+                ProviderError::new_with_context(
+                    Resources::Clear, format!(
+                        "The cluster '{}' does not exist and creation policy '{:?}' requires that it exist",
+                        cluster_name,
+                        creation_policy
+                    )
+                )
+            ),
+        CreationPolicy::Create  =>{
+            Ok((true, format!("Creation policy is '{:?}' and cluster '{}' does not exist: creating cluster", creation_policy, cluster_name)))
+        },
+        CreationPolicy::IfNotExists if !*cluster_exists => {
+            Ok((true, format!("Creation policy is '{:?}' and cluster '{}' does not exist: creating cluster", creation_policy, cluster_name)))
+        },
+        CreationPolicy::IfNotExists |
+        CreationPolicy::Never => {
+            Ok((false, format!("Creation policy is '{:?}' and cluster '{}' exists: not creating cluster", creation_policy, cluster_name)))
+        },
+    }
```
Should this go into agent/utils?

etungsten replied:
The logic/processing seems too specific to be generalized in an "utils" module. I was debating on whether to create a `cluster` module for this but that seems too generic of a name for a module, so having it here in lib.rs was where I landed.

ecpullen commented on bottlerocket/agents/src/tuf.rs:
```diff
@@ -6,7 +7,29 @@ use url::Url;
 
-pub(crate) fn download_target(
+/// Parse both TUF urls for downloading OVAs
+pub fn tuf_repo_urls(
```
Should this move to agent/utils?

etungsten replied:
There are other TUF-related functions here in the `tuf` module that are used to interact with a TUF repository.

ecpullen replied:
I meant the entire module. I think we can keep it here for now, though.

ecpullen reviewed (approved):
Just a couple questions.

etungsten commented:
Talked with @mjsterckx and we're ok with merging
